license_manager/service/service.go

```diff
@@ -2,36 +2,295 @@ package service
+		lic.Promote()
+	}
+}
+
+// Allocate allocates a license to the requesting invocation, or queues the
+// request if none are available. See the proto docstrings for more details.
 func (s *Service) Allocate(ctx context.Context, req *lmpb.AllocateRequest) (*lmpb.AllocateResponse, error) {
-	// If invocation is already allocated, return allocation
-	// Otherwise, get spot in queue
-	//   If request has an ID, find that spot in the queue or error
-	//   Otherwise, request goes at the back of the queue for that license
-	// Attempt to allocate available licenses to everyone in queue
-	// If not at the front of the queue, return queue spot
-	// (At the front of the queue)
-	// If license not available, return queue spot
-	// Allocate license and return allocation
-	return nil, status.Errorf(codes.Unimplemented, "Allocate() is not yet implemented")
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if len(req.GetLicenses()) != 1 {
+		return nil, status.Errorf(codes.InvalidArgument, "licenses must have exactly one license spec")
+	}
+	licenseType := formatLicenseType(req.GetLicenses()[0])
+	lic, ok := s.licenses[licenseType]
+	if !ok {
+		return nil, status.Errorf(codes.NotFound, "unknown license type: %q", licenseType)
+	}
+	invocationID := req.GetInvocationId()
+
+	if invocationID == "" {
+		// This is the first AllocationRequest by this invocation. Generate an ID
+		// and queue it.
+		var err error
+		invocationID, err = generateRandomID()
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to generate invocation_id: %v", err)
+		}
+		inv := &invocation{
+			ID:          invocationID,
+			Owner:       req.GetOwner(),
+			BuildTag:    req.GetBuildTag(),
+			LastCheckin: timeNow(),
+		}
+		lic.Enqueue(inv)
+
+		switch s.currentState {
+		case stateStarting:
+		case stateRunning:
+			lic.Promote()
+		default:
+			return nil, status.Errorf(codes.Internal, "unhandled state: %v", s.currentState)
+		}
+	}
+
+	// Invocation ID should now be known and either queued (due to the above
+	// insert, or from a previous request) or allocated (promoted from the queue
+	// by above, or asynchronously by the janitor).
+
+	if inv := lic.GetAllocated(invocationID); inv != nil {
+		// Invocation is allocated
+		inv.LastCheckin = timeNow()
+		return &lmpb.AllocateResponse{
+			ResponseType: &lmpb.AllocateResponse_LicenseAllocated{
+				LicenseAllocated: &lmpb.LicenseAllocated{
+					InvocationId:           invocationID,
+					LicenseRefreshDeadline: timestamppb.New(timeNow().Add(s.allocationRefreshDuration)),
+				},
+			},
+		}, nil
+	}
+	if inv := lic.GetQueued(invocationID); inv != nil {
+		// Invocation is queued
+		inv.LastCheckin = timeNow()
+		return &lmpb.AllocateResponse{
+			ResponseType: &lmpb.AllocateResponse_Queued{
+				Queued: &lmpb.Queued{
+					InvocationId: invocationID,
+					NextPollTime: timestamppb.New(timeNow().Add(s.queueRefreshDuration)),
+				},
+			},
+		}, nil
+	}
+	// Invocation is not allocated or queued
+	switch s.currentState {
```

Comment:
nit: I'd personally also use an if for readability:

```
if s.currentState != stateStarting {
  return nil, status.Errorf(...)
}

inv := &invocation{
...
}
lic.Enqueue(ivn)
...
```

Reply:
Done; removed all the switches.